pkg/kv/kvserver/batcheval/cmd_gc.go

```diff
@@ -184,10 +204,35 @@ func GC(
 		}
 	}
 
+	desc := cArgs.EvalCtx.Desc()
+
+	if rk := args.ClearRangeKey; rk != nil {
+		// Check if we are performing a fast path operation to try to remove all user
+		// key data from the range.
+		if rk.StartKeyTimestamp.IsEmpty() {
+			if !rk.StartKey.Equal(desc.StartKey.AsRawKey()) || !rk.EndKey.Equal(desc.EndKey.AsRawKey()) {
+				return result.Result{}, errors.Errorf("gc with clear range operation could only be used on the full range")
+			}
+
+			if err := storage.MVCCGarbageCollectWholeRange(ctx, readWriter, cArgs.Stats,
+				rk.StartKey, rk.EndKey, cArgs.EvalCtx.GetGCThreshold(),
+				cArgs.EvalCtx.GetMVCCStats()); err != nil {
+				return result.Result{}, err
+			}
+		} else {
+			// Otherwise garbage collect specified keys defined by clear range i.e. parts
+			// of the whole range containing no data.
+			if err := storage.MVCCGarbageCollectPointsWithClearRange(ctx, readWriter, cArgs.Stats,
+				rk.StartKey, rk.StartKey.Next(), rk.StartKeyTimestamp,
```

Comment:
Shouldn't this be `EndKey`? Why didn't tests catch this?

Reply:
This is bad. End to end test need to cover that. Randomized tests do this machinery by themselves. I'll add a test there.

Reply:
So I added variants to end to end test where it relies on full range deletion, point deletion and range with timestamp deletion methods. And it picks this bug. So all request variants should be covered. Unfortunately test needs to make assumptions that certain settings will cause certain modes of operation, but if this changes or heuristic changes test may become less useful.